definitions/npm/ramda_v0.26.x/flow_v0.104.x-/ramda_v0.26.x.js

```diff
@@ -1172,7 +1172,7 @@ declare module ramda {
 
   declare function transpose<T>(xs: Array<Array<T>>): Array<Array<T>>;
 
-  declare function uniq<T>(xs: Array<T>): Array<T>;
+  declare function uniq<A, As: $ReadOnlyArray<A>>(As): As;
```

Comment:
`$ReadOnlyArray` in, and `Array` out.

Reply:
I tried yesterday when working on the 'nth' factor, even though the output is read only,  I observe no problem when composing it using this definition, as ramda won't change any user data, I think all its function will make a copy of user input and operate on them.

  declare type __CurriedFunction1<A, R, AA: A> = (...r: [AA]) => R;
  declare type CurriedFunction1<A, R> = __CurriedFunction1<A, R, *>;

  declare function append<A, E>(A, $ReadOnlyArray<E>): Array<E|A>;
  declare function append<A, E>(A): CurriedFunction1<$ReadOnlyArray<E>, Array<E|A>>;
  
 declare function uniq<A, As: $ReadOnlyArray<A>>(As): As;
//test
const x:$ReadOnlyArray<number> = append(1, uniq([2, 2]))

On the other side, if we want to assign the result of uniq to a mutable array, namely, down casting that "should" cause an error, which you questioned below on the test. Or we want the output to not be read only because we want to perform some normal operations on it via it's native methods, or some other reason that I missed?

Reply:
The reason we want to use `Array` as an output is so we can still use the result in _another_ list that might have nothing to do with Ramda. Remember that there is no `$ReadOnlyArray` in JavaScript. It's a Flow idiocentricity that allows you to say "I promise I won't change this one", but has no assurances that the array will never change (this assurance is impossible).

To reiterate:
* `Array` is mutable.
* `$ReadOnlyArray` is mutable, just not _here_ (for relevant contexts of "here").
* There is no `$ReadOnlyArray` in the runtime. There is only `Array`.
* A frozen `Array` is not a `$ReadOnlyArray`.

Not being able to use an array you give to a Ramda function in something that doesn't use a universal `$ReadOnlyArray` _is a big problem_. Our changes we're putting together here should not make us lose any existing functionality/correctness.

Reply:
Thanks for the explanation, the desired change is made.